Master kickstart dispatcher anaconda integration

rvykydal opened the pull request:
Integrate kickstart distpatcher in Anaconda:

- Create Boss interface for early stage of anaconda where there is no event loop so we will call boss methods synchronously or poll the state (modules available).
- Boss (KickstartManager) asks modules for lists of handled elements and gives them partial kickstart (string).
- Mapping between module kickstart and original kickstart lines is handled by Boss (KickstartManager).
- Kickstart parsing in modules is not implemented, there is a dummy implementation returns no-error return value for now and it is possible to trigger parsing error by having PARSE_ERROR string in kickstart (eg "network --hostname=PARSE_ERROR")
- I'll ask Chris if we can modify KickstartParseError to contain lineno attribute. If not, we can be getting it from modified KickstartParser (KickstartParser for modules).

poncovka commented on pyanaconda/modules/boss/kickstart_manager.py:
```diff
@@ -0,0 +1,139 @@
+    pass
+
+
+class SplitKickstartUnknownSectionError(SplitKickstartError):
```
I am not sure, but maybe all DBus errors should be decorated.

poncovka commented on pyanaconda/modules/boss/kickstart_manager.py:
```diff
@@ -0,0 +1,139 @@
+        ksparser = SplitKickstartParser(handler, valid_sections=VALID_SECTIONS_ANACONDA)
+        try:
+            result = ksparser.split(path)
+        except KickstartParseError as e:
```
Actually, you can map the` KickstartParseError` and `KickstartError` to a DBus error if you want to.

poncovka commented on pyanaconda/modules/boss/kickstart_manager.py:
```diff
@@ -0,0 +1,139 @@
+            log.info("distribute kickstart: {} will get kickstart elements: {}".format(
+                observer.service_name, elements))
+
+            error_lineno, error_msg = observer.proxy.ConfigureWithKickstart(kickstart)
```
We might need to get the warnings are well, because in the strict mode we treat them as errors. But that can be extended in another issue.

rvykydal commented:
Updated with pylint fixes.

rvykydal commented:
jenkins, test this please

jkonecny12 commented on pyanaconda/startup_utils.py:
```diff
@@ -433,6 +462,8 @@ def parse_kickstart(options, addon_paths):
         ksdata = kickstart.parseKickstart(ks, options.ksstrict)
 
+        if "run_boss" in flags.cmdline:
+            distribute_kickstart_with_boss(ks)
```
I'm not sure if there is better place but this is side effect and it will be hard to find how the kickstart is distributed when it's here. I would rather see this call somewhere else.

rvykydal replied:
I think we need either to split the parse_kickstart function or do passing to boss in this function. I'd clean this up (parse_kickstart) when we know better what we want to do (for example parse only modules-unprocessed part in anaconda?). For now, we just want to be able to pass kickstart to modules to be able to work on their POCs/integration.
What about - to be more explicit - adding pass_to_boss flag to the parse_kickstart function? I'll add a patch doing that to the PR.

jkonecny12 commented on tests/pyanaconda_tests/kickstart_manager.py:
```diff
@@ -0,0 +1,298 @@
+from contextlib import contextmanager
+import os
+
+from pyanaconda.dbus.observer import DBusObjectObserver
```
Group `import`s and `from`s together.

rvykydal commented on pyanaconda/modules/boss/boss.py:
```diff
@@ -25,25 +25,69 @@
 
+@dbus_interface(DBUS_BOSS_ANACONDA_NAME)
+class AnacondaBossInterface(object):
+    """Temporary interface for anaconda.
```
I mean temporary interface for the transition stage of development where we are moving functionality from anaconda to boss and modules.

M4rtinK commented on pyanaconda/modules/boss/boss.py:
```diff
@@ -25,25 +25,69 @@
+        :returns: list of (Module service, (Line number, File name), Error message)
+                  tuples for each kickstart parsing error.
+        """
+        log.info("distirbuting kickstart")
```
typo, should be "distributing kickstart"

M4rtinK commented on pyanaconda/modules/boss/boss.py:
```diff
@@ -25,25 +25,69 @@
+        log.info("distirbuting kickstart")
+        errors = self._kickstart_manager.distribute()   # pylint: disable=no-member
+        if errors:
+            log.info("distributng kickstart errors: %s", errors)
```
also typo, should be: "distributing kickstart errors: %s"

rvykydal commented:
- Fixed typos.
- I am leaving the new Exceptions for now (@poncovka), but I am mapping specific child errors instead of the parent to dbus errors now. Also renamed one of the exceptions to tell what happened more correctly (can be seen in tests).
- I've added one more patch for @jkonecny12 (parse_kickstart in anaconda)
